drivers/gpio/gpio_cmsdk_ahb.c

```diff
@@ -269,146 +269,44 @@ static int gpio_cmsdk_ahb_init(struct device *dev)
-	.gpio_cc_dss = {.bus = CMSDK_AHB, .state = SOC_DEEPSLEEP,
-			.device = DT_CMSDK_AHB_GPIO2,},
-};
-
-static struct gpio_cmsdk_ahb_dev_data gpio_cmsdk_ahb_2_data;
-
-DEVICE_AND_API_INIT(gpio_cmsdk_ahb_2,
-		    CONFIG_GPIO_CMSDK_AHB_PORT2_DEV_NAME,
-		    gpio_cmsdk_ahb_init, &gpio_cmsdk_ahb_2_data,
-		    &gpio_cmsdk_ahb_2_cfg, POST_KERNEL,
-		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &gpio_cmsdk_ahb_drv_api_funcs);
-
-static void gpio_cmsdk_ahb_config_2(struct device *dev)
-{
-	IRQ_CONNECT(DT_IRQ_PORT2_ALL, CONFIG_GPIO_CMSDK_AHB_PORT2_IRQ_PRI,
-		    gpio_cmsdk_ahb_isr,
-		    DEVICE_GET(gpio_cmsdk_ahb_2), 0);
-	irq_enable(DT_IRQ_PORT2_ALL);
-}
-#endif /* CONFIG_GPIO_CMSDK_AHB_PORT2 */
-
-/* Port 3 */
-#ifdef CONFIG_GPIO_CMSDK_AHB_PORT3
-static void gpio_cmsdk_ahb_config_3(struct device *dev);
-
-static const struct gpio_cmsdk_ahb_cfg gpio_cmsdk_ahb_3_cfg = {
-	.common = {
-		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_PROP(3, ngpios)),
-	},
-	.port = ((volatile struct gpio_cmsdk_ahb *)DT_CMSDK_AHB_GPIO3),
-	.gpio_config_func = gpio_cmsdk_ahb_config_3,
-	.gpio_cc_as = {.bus = CMSDK_AHB, .state = SOC_ACTIVE,
-		       .device = DT_CMSDK_AHB_GPIO3,},
-	.gpio_cc_ss = {.bus = CMSDK_AHB, .state = SOC_SLEEP,
-		       .device = DT_CMSDK_AHB_GPIO3,},
-	.gpio_cc_dss = {.bus = CMSDK_AHB, .state = SOC_DEEPSLEEP,
-			.device = DT_CMSDK_AHB_GPIO3,},
-};
-
-static struct gpio_cmsdk_ahb_dev_data gpio_cmsdk_ahb_3_data;
-
-DEVICE_AND_API_INIT(gpio_cmsdk_ahb_3,
-		    CONFIG_GPIO_CMSDK_AHB_PORT3_DEV_NAME,
-		    gpio_cmsdk_ahb_init, &gpio_cmsdk_ahb_3_data,
-		    &gpio_cmsdk_ahb_3_cfg, POST_KERNEL,
-		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &gpio_cmsdk_ahb_drv_api_funcs);
+#define PORT_PIN_MASK(n) \
+	GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_PROP(n, ngpios))
+
+#define CMSDK_AHB_GPIO_DEVICE(n)							\
+	static void gpio_## n ##_cmsdk_config_func(struct device *dev);			\
+											\
+	static const struct gpio_cmsdk_ahb_cfg gpio_cmsdk_port_## n ##_config = {	\
+		.common = {								\
+			.port_pin_mask = PORT_PIN_MASK(n),				\
+		},									\
+		.port = ((volatile struct gpio_cmsdk_ahb *)DT_INST_REG_ADDR(n)),	\
+		.gpio_config_func = gpio_## n ##_cmsdk_config_func,			\
+		.gpio_cc_as = {.bus = CMSDK_AHB, .state = SOC_ACTIVE,			\
+			       .device = DT_INST_REG_ADDR(n),},				\
+		.gpio_cc_ss = {.bus = CMSDK_AHB, .state = SOC_SLEEP,			\
+			       .device = DT_INST_REG_ADDR(n),},				\
+		.gpio_cc_dss = {.bus = CMSDK_AHB, .state = SOC_DEEPSLEEP,		\
+				.device = DT_INST_REG_ADDR(n),},			\
+	};										\
+											\
+	static struct gpio_cmsdk_ahb_dev_data gpio_cmsdk_## n ##_data;			\
+											\
+	DEVICE_AND_API_INIT(gpio_cmsdk_port_## n,					\
+			    DT_INST_LABEL(n),						\
+			    gpio_cmsdk_ahb_init,					\
+			    &gpio_cmsdk_## n ##_data,					\
+			    &gpio_cmsdk_port_## n ##_config,				\
+			    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,		\
+			    &gpio_cmsdk_ahb_drv_api_funcs);				\
```

Comment:
I wonder if it would be better to use more consistent naming:
* gpio_N_cmsdk_config_func => gpio_cmsdk_port_N_config_func
* gpio_cmsdk_port_N_config (no change)
* gpio_cmsdk_N_data => gpio_cmsdk_port_N_data
* gpio_cmsdk_port_N (no change)